### ISSUE TITLE
Keep track of checksum per gem in the lockfile

### ISSUE THREAD

**Issue by coilysiren** [rubygems/rubygems#3379]
_From @Bertg on September 30, 2013 8:49_

More and more people are getting worried about the safety of gems and the ecosystem around it. There are some nice project on their way (signed gems etc...) but it might still be some time before these are main stream. And that might still not fix issues with git based gems etc.

What I'm suggesting here is not a 100% fail safe system. What I'd like to do here is add to Bundler's "The gems you run locally will be the gems on your server" guarantee. 

I'm suggesting to add a checksum to the Gemfile.lock file per gem. When creating the lockfile we create a checksum per gem and store it in the file.
On install, specifically when retrieving a gem from a foreign source, we again make the checksum and compare. 
This should add a significant amount of confidence that the source of the gem isn't compromised and that I'm running the exact same version on all machines. 

What do you think?

_Copied from original issue: bundler/bundler-features#27_

**Comment by pjmartorell** [comment]
What is the difference between this proposal and what Bundler already does? As far as I know, Bundler checks that the checksum of the gem you downloaded matches with the checksum stored in Rubygems. What am I missing? Is this feature only intended for git-based or foreign source gems?

**Comment by thbar** [comment]
@pjmartorell I would personally be interested to ensure the Bundler client compares the checksum stored in `Gemfile.lock` with what is on the server, to detect potential tampering on the server. This would give some guarantees that the locked package has not changed in content.

This is what other package managers do (e.g. Hex in Elixir, [introduced 4 years ago](https://github.com/hexpm/hex/blob/master/CHANGELOG.md#v0120-2016-05-15)).

**Comment by Bertg** [comment]
@pjmartorell In my suggestion (if I remember correctly, it's been some time) Bundler should store the checksums of the gems in the lock file for future reference.
This protects us against future tampering of gems and the services that provide them (man in the middle attacks, compromised servers, etc...) as we store the checksums in the lock file.

**Comment by pjmartorell** [comment]
As far as I know, in the case of RubyGems gems, Bundler already checks that the checksum of the downloaded gem matches with the checksum from RubyGems, so apparently there is no risk of tampering except RubyGems is compromised.

**Comment by Bertg** [comment]
@pjmartorell This is not about checking the current download with the current listing. This is about validating that the download of "2 weeks ago" on an other system is the same download it is today.

The current lock file does not contain ANY checksum information, so verification through time is not possible.
So it is also impossible to tell if the gem provider got compromised in the mean time.

**Comment by thbar** [comment]
> except RubyGems is compromised

@pjmartorell yes - and this is one scenario that I would definitely like to see covered.

**Comment by Bertg** [comment]
> except RubyGems is compromised

Driving the point home... 

It's a lot more than just RubyGems, it can be any source (GitHub, GitLab, a private repo)
And not just "now". If the lock file contains shas it can also indicate that a source was compromised in the past.
And not only the source. It can also give indication that the previous machine (owned by an other developer for instance) that updated the lock file was compromised. 

I also want to make clear that the ideal solution here should not only work for official gems; but any remote source including direct git commit references.

**Comment by bf4** [comment]
I believe this may be closed as completed

**Comment by deivid-rodriguez** [comment]
Could you clarify @bf4? We don't keep checksum information in the lockfile at the moment which I believe is what this issue is suggesting.

**Comment by bf4** [comment]
if the intention is to have the checksum available, I just mean it is. If the intention is to have it be automated, I don't believe it is.  Putting in the Gemfile.lock might possibly be useful for some 3rd party tool to validate against what's known via rubygems.org.  So, depends if the issue is 'keep track of checksum' vs 'keep a checksum in the Gemfile.lock' or 'have gem command which validates checksums'. If it's just 'keep track of checksum', I believe it's done. That's all I mean.

**Comment by Bertg** [comment]
Given the lates hype around "dependency confusion" https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610 the priority of this issue should be reconsidered?

**Comment by juanibiapina** [comment]
This could also be useful for packaging ruby applications with Nix.

**Comment by mperham** [comment]
I'll echo Bertg here. Given the rise in supply chain attacks, should this be revisited?

**Comment by indirect** [comment]
Checksums are good, checksums guard against (mostly) network transmission issues, checksums absolutely do not help at all with supply chain attacks, because by definition a supply chain attack means shipping a legitimate release, whose checksum would match, but contain malware.

**Comment by mperham** [comment]
@indirect I'm thinking of the (possibly miniscule?) risk that a bad actor can modify a .gem after initial publish by breaking into rubygems infrastructure.

**Comment by indirect** [comment]
The bad news is if they can break into rubygems infrastructure, they can also update the checksum to match :/

**Comment by indirect** [comment]
To be clear: I am strongly in favor of checksums, defense in depth, network errors, etc etc. But it doesn't solve the security issue.

**Comment by mperham** [comment]
The checksum is local to each  app, in Gemfile.lock. The attacker can't update that.

**Comment by mperham** [comment]
This would be the TOFU principle. Bundle install would calculate the hash on first download and save it in the lockfile. Future deploys would verify the hash.

**Comment by deivid-rodriguez** [comment]
I understand the motivations of this feature request and I'm happy if this feature gets added to `bundler` :). Having it would've indeed prevented dependency confusion issues, but since those are fixed now, it doesn't seem like a motivation to implement it anymore. It would still be helpful to prevent other potential issues where a source gets compromised.

**Comment by indirect** [comment]
Depending on how they are implemented, checksums can be attacked in different ways: if checksums are calculated by the client on first install, they do not guard against network errors or supply chain attacks during the first install. If checksums are provided by the server, they do not guard against compromise of the server providing the checksums. We would, ideally, like to eventually have support for both kinds of checksum-checking. 👍🏻

**Comment by deivid-rodriguez** [comment]
Yep, I think this is about checksum checking on the client.

Basically, ensure via lockfile that once you get a specific copy of a gem installed, the checksum of what you installed is recorded, so that the exact same thing gets installed on other machines using that lockfile. So if you get a compromised version on first install, bundler will refuse to go on on other machines unless they get that compromised version too :laughing:.

**Comment by thbar** [comment]
I would find interesting to look at what is done elsewhere, and in particular on [Hex.pm](https://hex.pm), the package manager for Elixir and Erlang, which stores the checksums in a registry which is signed by a private key.

I have contacted @wojtekmach from the [hexpm](https://github.com/hexpm) team to get more information on this evergoing work in progress.

He shared that (and will maybe comment here if he will): 

> in a nutshell the reason we have a checksum in the lock is to ensure the package tarball hasn't been tampered with. We compare the downloaded tarball checksum against the checksum in the registry. We can trust the checksum in the registry because the registry is signed with a private key. We verify the signature against a public key that Hex ships with. You can read a bit about that aspect here https://blog.voltone.net/post/25

They have plans to document what is done in this regard with more details [here](https://github.com/hexpm/hexpm/issues/630), I suppose they will detail the threat model. This will probably describe the notions of inner and outer checkums that [can be found](https://github.com/hexpm/hexpm/search?q=checksum) in the codebase, too.

Also there are interesting comments by Eric from the Hex team too [here](https://stackoverflow.com/a/48583473).

Finally, there are some useful bits in [this blog post](https://yiming.dev/blog/2020/06/29/making-audit-logs-on-hex-pm-public/), which could give interesting ideas.

Having a form of immutability and/or tampering detection in the data somewhere would definitely help. Having a design that would protect at least a bit in case of partial compromise of the rubygems infrastructure would be sweet to pursue (what a honeypot).

As a matter of personal opinion, like @deivid-rodriguez one thing I would find useful would be to have checkums in the Gemfile.lock, which would be compared against at installation time, ideally before running any script. While this would not cover against picking a vulnerable version in the first place, it would alert us at deployment time, that something has been tampered, and this would already provide a good first step.

**Comment by mensfeld** [comment]
> I understand the motivations of this feature request and I'm happy if this feature gets added to bundler :). Having it would've indeed prevented dependency confusion issues, but since those are fixed now, it doesn't seem like a motivation to implement it anymore. It would still be helpful to prevent other potential issues where a source gets compromised.

Well first of all: one vector of attack is fixed. We do not know if there's another one.

Also as people before mentioned: data in S3 can be replaced, caches can be poisoned. This is a great feature.

On top of that, I do recall a case (I think I discussed it with @sonalkr132 but I may be wrong) where RubyGems would serve a different gem due to some mistake.

All of those would be prevented that way.

It won't fix this problem though: https://www.whitesourcesoftware.com/resources/blog/lockfiles-security/ - but this can be mitigated with --frozen and as long as frozen would anyhow enforce checksums instead of "rebuilding" them, it should be fine.

**Comment by deivid-rodriguez** [comment]
In case it was not not clear by previous messages, I'm strongly in favor of adding this.

**Comment by mensfeld** [comment]
In regards to performance that should not be a problem. Majority of packages are small. Computing all the 446 gems checksums on my average PC took (107MB) took less than 0.2s:

```
time find -type f -exec md5sum "{}" + > checklist.chk

real	0m0.177s
user	0m0.156s
sys	0m0.020s
```

even with sha256:

```
time find -type f -exec sha256sum "{}" + > checklist.chk
real	0m0.520s
user	0m0.503s
sys	0m0.013s
```

on top of that this could be cached alongside the other metadata so only the first run would need to do it. Unless we want to also ensure consistency of packages present after install to make sure of their post instal consistency but this would be a huge challenge and I think it is not what we want to achieve here.

**Comment by Bertg** [comment]
> Unless we want to also ensure consistency of packages present after install to make sure of their post instal consistency

Honestly, I think this would be preferable. An attack could also happen agains the local copies of any gem, or the cached values for that matter. 

Adding the caching logic would be more work, and the performance is acceptable. So I woud just do the full check each time it is possible.

**Comment by jsierles** [comment]
I'd also like this for Nix support. Tools like [dream2nix](https://github.com/nix-community/dream2nix) would benefit greatly from having this information up front.

**Comment by kuahyeow** [comment]
I am interested in potentially contributing, though I have few questions about Bundler lockfiles.

For context, we at GitLab are working on hardening the checksum process by tracking the checksums stored in a committed file. A draft is present at https://gitlab.com/gitlab-org/gitlab/-/merge_requests/92633. For the purposes of the initial solution, we are monkey-patching Bundler, and using a separate file. If this proves out well, we will push a PR up to Bundler

I understand there is a desire to create a backwards compatible section in Bundler lockfile to store the checksums. My major concern is platforms.

1. How does Gemfile.lock `PLATFORMS` work ? Does Bundler calculate the relevant local platform that matches the list of `PLATFORMS` and then install the gem for that platform ?
2. Is `ruby` platform a special case in any way ?
3. How would storing checksums work for multi-platform gems ? To take `nokogiri` as an example. I specific `ruby` as the only `PLATFORMS` in Gemfile.lock. Running `bundle install` on a Mac M1 will install `nokogiri-1.13.8-arm64-darwin.gem`. Someone else on a linux machine will install `nokogiri-1.13.-x86_64-linux.gem`. We might have to take checksums for all platforms from the Rubygems server instead to address this. Would that work ?

/cc @deivid-rodriguez

**Comment by deivid-rodriguez** [comment]
Hello @kuahyeow!

1) How does Gemfile.lock PLATFORMS work ? Does Bundler calculate the relevant local platform that matches the list of PLATFORMS and then install the gem for that platform ?

Yes. Multiplatform Gemfiles keep a resolution valid for multiple platforms. When `bundle install` or `bundle exec` are run, Bundler will select only the gems relevant for the current platform.

2) Yes, "ruby" is special (and things are still a bit of a mess here). Before we started locking the specific platform in the lockfile, Bundler would just use "ruby" in the PLATFORMS section of the lockfile, and would not lock platform specific gems in the lockfile, just "generic versions". So Bundler will for example see "nokogori-1.13" in the lockfile, and right before install decide that it needs to install 1.13.8-arm64-darwin.gem specifically. But on a different platform, say linux, it will install `1.13.8-x86_64-linux` instead. Even if the lockfile is the same. This special behaviour is preserved these days for backwards compatibility (only if there's no more specific platforms than "ruby" in the lockfile). But should be eventually removed, and using "ruby" should always mean that only gems without platform specific variants should be installed.

3) I think for now we should only store the checksum for "modern" lockfiles that store specific platforms, since we won't need anything special, just lock the checksum of each gem included in the lockfile.

**Comment by kuahyeow** [comment]
Thanks @deivid-rodriguez  !

> I think for now we should only store the checksum for "modern" lockfiles that store specific platforms, since we won't need anything special, just lock the checksum of each gem included in the lockfile.

I have opened a draft PR in https://github.com/rubygems/rubygems/pull/5808. It's DRAFT because there is a lot of work in adding a new lockfile section, and I want to get some early feedback.

**Comment by sdwolfz** [comment]
@deivid-rodriguez @kuahyeow Can you please clarify one thing?

Say I develop on a macbook, installing a platform specific gem, that means the lockfile will contain a hash for `arm64-darwin` gem only. I commit and deploy to a Linux server, so when `bundle install` runs, it will only check the hash if it installs `arm64-darwin` again. But since it's actually `x86_64-linux` now, it won't actually check the hash, but will update the `Gemfile.lock` by appending the new value.

:point_up: Is my understanding of this correct? If yes then it looks like developing on a different arch than your servers will basically invalidate this feature, plus `Genfile.lock` will start seeing edits in paces where previously you could assume it's not changing (like in CI, which can trigger unwanted `git diff --exit-code` failures).

Not sure what to propose as a fix for this, downloading all platform flavors of the gem and storing all hashes is not ideal, and spiking over hash checks is something I'd like to avoid. How do other package managers handle this?

**Comment by Bertg** [comment]
Just dropping in my 2 cents, this is way over my pay grade, but I was the one originally suggesting this feature... 

> Not sure what to propose as a fix for this, downloading all platform flavors of the gem and storing all hashes is not ideal, and skipping over hash checks is something I'd like to avoid. 

I think, as the purpose here is to increase safety in case of a supply chain attack, that each "flavour" that could be resolved, should be covered by this feature. If that means downloading, creating and storing hashes, so be it. 

I think the impact will not be that huge, as not that many gems have platform specific packages.

I think, in a later stage, we could allow the developer to choose what possible platforms this feature should cover, as a way of optimising the process.

**Comment by deivid-rodriguez** [comment]
I think this connects with #5700. The idea is that you should generate your `Gemfile.lock` for your deployment platform before you actually deploy. This is already like that, #5700 is just trying to make that more user friendly.

With that in mind, I think you should be able to generate your `Gemfile.lock` file for both `arm64-darwin` and `x86_64-linux` locally without installing anything and get all checksums recorded by using the checksum information provided by the RubyGems API. Then, at install time, you can compare the checksum recorded in the lockfile with the actual checksum of the `.gem` package, regardless of the platform. Bundler already does this now, but the checksum information lives in the compact_index cache (somewhere at `~/.bundle/cache/compact_index/`), not in the lockfile. So it's not part of the application and it's lost when the cache is removed.

**Comment by kuahyeow** [comment]
I agree with https://github.com/rubygems/rubygems/issues/3379#issuecomment-1203638023.

If we take https://github.com/rubygems/rubygems/pull/5808 together with `--frozen`, then it will be unpleasant experience generating a lockfile on  a development machine (macOS), which will then fail on CI (different platform like linux). 

So https://github.com/rubygems/rubygems/pull/5700 could help in this a lot ! 

But also, we can take the problem further. Lockfiles are frozen too on `--deployment` so is it possible/advisable to make lockfiles generated on one platform work for all platforms listed under `PLATFORMS` ?

**Comment by deivid-rodriguez** [comment]
> If we take https://github.com/rubygems/rubygems/pull/5808 together with --frozen, then it will be an unpleasant experience generating a lockfile on a development machine (e.g. macOS), which will then fail on CI (different platform like linux). So https://github.com/rubygems/rubygems/pull/5700 could help in this a lot !

One of my main goals for this year is to move forward both #5700 and #5808 😃.

> But also, we can take the problem further. Lockfiles are frozen too on --deployment so is it possible/advisable to make lockfiles generated on one platform work for all platforms listed under PLATFORMS ?

Note that `deployment` just means `frozen + path=vendor/bundler`. We're actually [considering](https://github.com/rubygems/rubygems/issues/4534) making it just an alias of `frozen`. So everything that applies to `frozen` should also apply to `deployment`. Does that change anything? I'm not sure I followed the above.

Maybe the title of #5700 ("Make lockfiles generated on MacOS also work on Linux by default") is a bit misleading. There's really no 100% guarantee that anything you have only tested on MacOs will work on Linux, everyone should still try/test anything in the deployment platform before deploying it 😅. My intention is only to make default lockfiles "Bundler friendly" on multiple platforms, even in frozen mode, regardless of which platform is used when generating them.

**Comment by kuahyeow** [comment]
Thanks @deivid-rodriguez  !

> Note that `deployment` just means `frozen + path=vendor/bundler`. We're actually [considering](https://github.com/rubygems/rubygems/issues/4534) making it just an alias of `frozen`. So everything that applies to `frozen` should also apply to `deployment`. Does that change anything? I'm not sure I followed the above.
> 
> Maybe the title of #5700 ("Make lockfiles generated on MacOS also work on Linux by default") is a bit misleading. There's really no 100% guarantee that anything you have only tested on MacOs will work on Linux, everyone should still try/test anything in the deployment platform before deploying it 😅. My intention is only to make default lockfiles "Bundler friendly" on multiple platforms, even in frozen mode, regardless of which platform is used when generating them.

To clarify, my question..say we have the following:

```
PLATFORMS
  aarch64-linux
  arm64-darwin
  java
  x86_64-linux
  x64-mingw-ucrt
```

Would the lockfile lock the specs for all platforms as a result ?

**Comment by deivid-rodriguez** [comment]
Yes, it should lock the most specific set of specs that works for all platforms. So, for example, if a gems has `x86_64-linux` precompiled version, but no `aarch64-linux` precompiled version, it should lock both the `x86_64-linux` precompiled version and also the generic version (to be used on `aarch64-linux`). I believe it's also how it works today.

**Comment by segiddins** [comment]
This is close to being done in https://github.com/rubygems/rubygems/pull/6374

**Comment by sdwolfz** [comment]
Sorry but I'm confused, #6374 has been merged, but how does one actually use this? Or is it not completed/released yet?

**Comment by simi** [comment]
Hello @sdwolfz, per my understanding it is merged, but it is not enabled (exposed) to users yet (hence no changelog entry). For now (according to https://github.com/rubygems/rubygems/pull/7217) you need to manually opt-in by adding empty CHECKSUMS section into Gemfile.lock manually. Commands to opt-in (and potential switch to enable by default) will follow in following releases once this feature will be considered stable enough.

**Comment by mensfeld** [comment]
@simi do we have an ETA when this moves from experimental to stable?

**Comment by deivid-rodriguez** [comment]
I'm working on adding a better way to enable this feature. No ETA, but I hope to have something for end of year's release.

**Comment by mensfeld** [comment]
@deivid-rodriguez amazing. I've been using and testing it for a while now and did not encounter any issues FYI.

**Comment by Earlopain** [comment]
https://github.com/rubygems/rubygems/pull/8214 has been merged, which means you can do `--add-checksums` in future versions.

**Comment by deivid-rodriguez** [comment]
Yes, we will release these ways of enabling the feature in Bundler 2.6, and call that a beta release of the lockfile checksums feature 👍.

**Comment by deivid-rodriguez** [comment]
This has been released with [Bundler 2.6](https://bundler.io/blog/2024/12/19/bundler-v2-6.html), thanks to everyone who worked on this!
